VocaDbWeb.Core/Controllers/UserController.cs

```diff
@@ -0,0 +1,149 @@
+		private readonly ArtistQueries _artistQueries;
+		private readonly ArtistService _artistService;
+		private readonly VdbConfigManager _config;
+		private UserQueries Data { get; set; }
+		private readonly IPRuleManager _ipRuleManager;
+		private readonly LoginManager _loginManager;
+		private readonly MarkdownParser _markdownParser;
+		private readonly UserMessageQueries _messageQueries;
+		private readonly OtherService _otherService;
+		private readonly IRepository _repository;
+		private UserService Service { get; set; }
+
+		public UserController(UserService service, UserQueries data, ArtistService artistService, ArtistQueries artistQueries, OtherService otherService,
+			IRepository repository,
+			UserMessageQueries messageQueries, IPRuleManager ipRuleManager, VdbConfigManager config, MarkdownParser markdownParser, ActivityEntryQueries activityEntryQueries, LoginManager loginManager)
+		{
+			Service = service;
+			Data = data;
+			_activityEntryQueries = activityEntryQueries;
+			_artistQueries = artistQueries;
+			_artistService = artistService;
+			_repository = repository;
+			_otherService = otherService;
+			_messageQueries = messageQueries;
+			_ipRuleManager = ipRuleManager;
+			_config = config;
+			_markdownParser = markdownParser;
+			_loginManager = loginManager;
+		}
+
+		[RestrictBannedIP]
+		public ActionResult Login(string returnUrl = null)
+		{
+			RestoreErrorsFromTempData();
+
+			return View(new LoginModel(returnUrl, false));
+		}
+
+		[RestrictBannedIP]
+		public PartialViewResult LoginForm(string returnUrl)
+		{
+			return PartialView("Login", new LoginModel(returnUrl, false));
+		}
+
+		[HttpPost]
+		[RestrictBannedIP]
+		public async Task<ActionResult> Login(LoginModel model)
+		{
+			if (ModelState.IsValid)
+			{
+				var host = WebHelper.GetRealHost(Request);
+				var culture = WebHelper.GetInterfaceCultureName(Request);
+				var result = Data.CheckAuthentication(model.UserName, model.Password, host, culture, true);
+
+				if (!result.IsOk)
+				{
+					ModelState.AddModelError("", ViewRes.User.LoginStrings.WrongPassword);
+
+					if (result.Error == LoginError.AccountPoisoned)
+						_ipRuleManager.AddTempBannedIP(host, "Account poisoned");
+				}
+				else
+				{
+					var user = result.User;
+
+					TempData.SetSuccessMessage(string.Format(ViewRes.User.LoginStrings.Welcome, user.Name));
+					var claims = new List<Claim>
+					{
+						new Claim(ClaimTypes.Name, user.Name),
+					};
+					var claimsIdentity = new ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationScheme);
+					var authProperties = new AuthenticationProperties
+					{
+						IsPersistent = model.KeepLoggedIn,
+					};
+					await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme, new ClaimsPrincipal(claimsIdentity), authProperties);
+
+					string redirectUrl = null;
+					// TODO: implement
+					/*try
+					{
+						redirectUrl = FormsAuthentication.GetRedirectUrl(model.UserName, true);

```

Comment:
If you try to access a page that requires login, such as Admin page, you get redirected to login page and the admin page URL will be in GetRedirectUrl, so after login you'd be redirected to the Admin page, where you wanted to get in the first place. This is not terribly important on VocaDB, because on normal pages the login happens in a popup dialog so redirecting is not needed. It's probably ok to leave it out initially, but should be reimplemented eventually.

Reply:
It seems that it's already done by the `ReturnUrl` query string?

Reply:
I think ReturnUrl is added by VocaDB to override the ASP.NET redirectUrl? But most likely one of them is not needed.